Revisions following test run (#560)
merely-useful/py-rse: chapters/testing.Rmd

expected_result = Counter(risk_poem_counts)
```

> We then generate the actual result by calling `countwords`,

Reviewer: ```suggestion
We then generate the actual result by calling `count_words`,
```

In order to create this text fixture,
we need a way to generate random words.

Fortunately, a Python library called 
[`RandomWordGenerator`](https://github.com/AbhishekSalian/Random-Word-Generator) 
exists to do just that.
We can install it using [`pip`][pip], the Python Package Installer:

```bash
$ pip install Random-Word-Generator
```

Borrowing from the word count distribution we created for `test_alpha`,
we can use the following code to create a text file full of random words
with a frequency distribution that corresponds to an \(\alpha\) of approximately 1.0:

```python
import numpy as np
from RandomWordGenerator import RandomWord

max_freq = 600
word_counts = np.floor(max_freq / np.arange(1, max_freq + 1))
rw = RandomWord()
random_words = rw.getList(num_of_words=max_freq)
writer = open('test_data/random_words.txt', 'w')
for index in range(max_freq):
    count = int(word_counts[index])
    word_sequence = f'{random_words[index]} ' * count
    writer.write(word_sequence + '\n')
writer.close()

> ```

Reviewer: It might be worth showing the abbreviated output from 
```bash
$ cat test_data/random_words.txt
```
just to demonstrate that it is a file full of random words?

In #567 I've done the same earlier in this chapter to show that `word_counts` is a numpy array full of synthetic word count data.

```

```text
bin/countwords.py      20      7    65%   25-26, 30-37
bin/plotcounts.py      58     37    36%   48-55, 75-77, 81-82, 
                                          86-118, 121-139
bin/test_zipfs.py      31      0   100%
bin/utilities.py        8      5    38%   18-22
-------------------------------------------------

> TOTAL                 117     49    58%

Reviewer: Thanks for updating the output from `$ coverage report -m`. Would you also be able to update Figure 11.1? It's a screenshot of the `$ coverage html` output for `countwords.py` and it currently shows a very old version of that script.

```

```text
bin/countwords.py      20      7    65%   25-26, 30-38
bin/plotcounts.py      58     37    36%   48-55, 75-77, 82-83, 
                                          88-118, 122-140
bin/test_zipfs.py      31      0   100%
bin/utilities.py        8      5    38%   18-22
-------------------------------------------------

> TOTAL                 117     49    58%

Reviewer: The figure corresponding to this coverage report (`figures/testing/python-coverage.png`) used to have a white background, while the new one has a black background. I'm wondering if in the printed book a white background would look better? (Don't worry if it's not easy for you to regenerate the new figure with a white background.)

Author: Done! Thanks for reminding me; I had to switch my OS preference back from dark (I prefer my computer to reflect my soul)

This output confirms that all lines relating to word counting were tested,
but not any of the lines related to argument handling or I/O.

> > **Commit our coverage?**

Reviewer: ```suggestion
> **Commit our Coverage?**
```

so that it knows which of its standard \gref{virtual machines}{virtual_machine} to use\index{continuous integration!use of virtual machines}
as a starting point for the project.
The `python` key specifies the version or versions of Python to use,
while the `install` key indicates the name of the file (`requirements.txt`)
listing the libraries that need to be installed.
The `script` key lists the commands to run---in this case, `pytest`---

> which in turn executes our project scripts.

Reviewer: ```suggestion
which in turn executes our project scripts
```

> (i.e. `test_zipfs.py`, `plotcounts.py`, `countwords.py` and `utilities.py`)

Reviewer: ```suggestion
(i.e. `test_zipfs.py`, `plotcounts.py`, `countwords.py` and `utilities.py`).
```

These scripts import a number of packages that don't come with the
[Python Standard Library][python-standard-library],
so these are what comprises `requirements.txt`:

```text
numpy
pandas
matplotlib
scipy
pytest
pyyaml
```

Be sure to save both `requirements.txt` and `.travis.yml`
to your project's root directory,

> commit to your repository,a

Reviewer: ```suggestion
commit to your repository,
```

$ git commit -m "Adding requirements"
```
```text
[master 4917a19] Adding requirements
 2 files changed, 9 insertions(+)
 create mode 100644 requirements.txt
```
```bash
$ git push origin master
```
```text
Enumerating objects: 6, done.
Counting objects: 100% (6/6), done.
Delta compression using up to 4 threads
Compressing objects: 100% (3/3), done.
Writing objects: 100% (4/4), 445 bytes | 74.00 KiB/s, done.
Total 4 (delta 1), reused 0 (delta 0), pack-reused 0
remote: Resolving deltas: 100% (1/1), completed with 1 local object.
To https://github.com/amira-khan/zipf.git
   46cd2d8..4917a19  master -> master
```

> Travis CI automatically runs again.

Reviewer: Sorry if you're already planning to do this, but you'll need to re-create `figures/testing/travis-build-fail.png` and `figures/testing/travis-build-fail.png`, because those are currently screen shots from my GitHub account as opposed to Amira's.

Author: Yup, I'm on this!
